removal of klog from validaters

Klaven opened the pull request:
Removed klog from validators. All errors and warnings are returned as such.

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #
xref: https://github.com/kubernetes/kubeadm/issues/1738
this only addresses the klog part, a follow up with the errors will be later.

**Special notes for your reviewer**:
Again, at this point i'm just looking for feedback, not really happy with it myself.
**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

yastij commented:
/assign @yastij @neolit123 @rosti

neolit123 commented on test/e2e_node/e2e_node_suite_test.go:
```diff
@@ -125,7 +125,7 @@ func TestE2eNode(t *testing.T) {
 			}
 		}
-		if _, err := system.ValidateSpec(*spec, framework.TestContext.ContainerRuntime); err != nil {
+		if _, _, err := system.ValidateSpec(*spec, framework.TestContext.ContainerRuntime); err != nil {
```
we'd need to handle the LogMessage output here too.

neolit123 commented:
thanks @Klaven 
i will let @rosti and @yastij have a look too.

Klaven commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -560,6 +561,19 @@ func (sysver SystemVerificationCheck) Check() (warnings, errorList []error) {
 	return warns, nil
 }
 
+func MessageLogger(msgs []system.LogMessage) {
```
this func is a way we could use the new `LogMessage` struct in kubeadm. but I figure we need a better spot to put then it and was unsure.

Klaven commented on cmd/kubeadm/app/util/system/logutils.go:
```diff
@@ -0,0 +1,60 @@
+package system
+
+// LogMessage is a struct for logs
+type LogMessage struct {
```
writing this entire file made me die a little inside. It felt VERY much like recreate a wheel that has been recreated 100 times. That being said I could not think of a better alternative last night.

Klaven commented on cmd/kubeadm/app/util/system/logutils.go:
```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
if we HAVE to keep this file (I really hope the answer is no) then I don't know where to put it. I stuck it in with the validators.... but in theory anything that gets removed from k/k should use it to return logs? right? what do you all think?

Klaven reviewed (commented):
Here is my review and the reason it is in WIP (other then some failing tests! i'll fix those then I have time this weekend) Thanks for suggestions/comments!

yastij commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -542,13 +542,14 @@ func (sysver SystemVerificationCheck) Check() (warnings, errorList []error) {
 	// Run all validators
 	for _, v := range validators {
-		warn, err := v.Validate(system.DefaultSysSpec)
+		logs, warn, err := v.Validate(system.DefaultSysSpec)
```
I would rather see this as `info, warn, err`

yastij commented on cmd/kubeadm/app/util/system/logutils.go:
```diff
@@ -0,0 +1,60 @@
+package system
+
+// LogMessage is a struct for logs
+type LogMessage struct {
```
I'm not sure we should adopt such thing. we should return array of strings for info, an array or one warn, and an err in case of error, the caller should decide of the verbosity to use

Klaven replied:
I don't think so. there are 3 different things we return.
Log messages that can be info/warn/error ect
None critical Errors
Critical Errors,
I originally thought just strings but it would leave a lot of info out. and we would no longer pass the errors out anymore, just some strings.

yastij replied:
I still think we should aggregate errs and logs and let the caller output them cc @neolit123 @rosti

Klaven replied:
Ok. I can do whatever you all think is best! If that is what you all want it's what I will do! If you read my own comments I am not a fan of this entire file. I just did not know of a way without just returning a flat string. and I thought that the caller lost a lot of info with a method like that. I would also be up for a zoom chat if you all wanted.

yastij replied:
I'd like to hear from other folks before we make any decision

Klaven replied:
:+1: this is why I posted it in [WIP] as I stated I don't like it how it is and wanted feedback!  Thanks for the comments! I'm just letting you know why I did not initially go with the `[]string` method :)

neolit123 replied:
@Klaven @yastij 
we should have probably discussed this a little or clarified in the related ticket.
we need a logging shim and we should not define custom objects and additional functional returns.

the wheel that was already invented here is called logr:
https://github.com/go-logr/logr/blob/master/logr.go

and klogr:
https://github.com/kubernetes/klog/blob/master/klogr/klogr.go

one way of approaching the problem is:

- create new klogr instances on the side of the system validator callers using `New()`.
- pass these instances as `logr.Logger` to the validators methods and constructors.
- use `mylog.Info` on the side of the validators.

but this forces the users of the validators to implement logr and logr is opinionated and complicated.

our alternative way here (that i prefer) is to just a pass logging shim from the validator callers to the validators:

```
// this is defined on the side of the validators
// we don't need more than Info* for now
type Logger interface {
  Info(args ...interface{})
  Infof(format string, args ...interface{})
}
```

so on the side of the callers one would define:
```
type myLogger struct {}

// the caller chooses the verbosity level, in this case 1
func (myLogger) Info(args ...interface{}) {
  klog.V(1).Info(args...)
}

func (myLogger) Infof(format string, args ...interface{}) {
  klog.V(1).Infof(format, args...)
}

...
myLog := &myLogger{}
// and pass this this to the validators as `system.Logger`
```

alternatively instead of passing this instance to all validator constructors etc.

the `system` package can have:
```
var logger Logger
var mutex sync.Mutex

func SetLogger(log Logger) {
  mutex.Lock()
  defer mutex.Unlock()
  logger = log
}
```

and then the validator code can call `logger.Infof`, which should panic if the consumer of the system validators has not called `SetLogger(...)`

yastij replied:
I'm fine with this, as long as we keep the interface in sync (although I don't expect it to change that much over time)

Klaven replied:
@neolit123 I like the shim, originally when I thought of the shim it sounded like a lot of reinventing, aka, `info`, `warn` and `error` as well as the ability to set verbosity. meaning everyone that tried to call these function would have to go though a bit of work to implement. But if I can just assume that all things are the same (and currently are NOT, we log with `info` and `error`) then this is in fact simple. I don't think the `SetLogger` is awesome just because it could be frustrating trying to use a function that requires some state set before hand that might not be self evident.

Klaven replied:
I also like `logr` :thinking: and probably would have gone down this path last night had I seen it. would it be possible for me to create a `klog` `logr` shim, so if you wanted to use `klog` it was already created and easy? and then it would just be anyone that wanted to log without klog that has to go though the painful stuff? I feel like just logging everything as `V(1).Info` sucks a little.

Klaven replied:
oh. klogr already is that shim :) sorry. missed that point. is there a reason you don't like this solution? anyone not using `klog` would need to implement it but is that a big problem?

yastij commented on cmd/kubeadm/app/util/system/package_validator.go:
```diff
@@ -127,7 +127,8 @@ func (validator *packageValidator) validate(packageSpecs []PackageSpec, manager
 		version, err := manager.getPackageVersion(packageName)
 		if err != nil {
-			klog.V(1).Infof("Failed to get the version for the package %q: %s\n", packageName, err)
+			logs = append(logs, logInfo(fmt.Sprintf("Failed to get the version for the package %q: %s\n", packageName, err), 1))
```
caller should decide about the verbosity (see above)

Klaven replied:
so all logs generated have the same verbosity? I'm not sure we want this.

yastij replied:
we have only `V(1)` here, also this would considerably simplify things

Klaven replied:
yes. we have a info and an error log :) but if just strings we would lose this.

Klaven replied:
I mean. I can change it to just strings for everything, []string for logs []string for warning errors and []string for error errors... but that feels bad to me. (and keep in mind I was not touching errors this PR. just logs. Errors' next PR.

Klaven commented on cmd/kubeadm/app/util/system/kernel_validator.go:
```diff
@@ -75,7 +74,8 @@ func (k *KernelValidator) Validate(spec SysSpec) (error, error) {
 	// only validate kernel config when necessary (currently no kernel config for windows)
 	if len(spec.KernelSpec.Required) > 0 || len(spec.KernelSpec.Forbidden) > 0 || len(spec.KernelSpec.Optional) > 0 {
-		errs = append(errs, k.validateKernelConfig(spec.KernelSpec))
+		err := k.validateKernelConfig(spec.KernelSpec)
```
lol. not sure why this got changed. will change it back in the morning. should be no change here.

rosti replied:
Can we revert the changes here? This does not affect the patch, but it's not relevant for the change.

rosti reviewed (changes requested):
Thanks for picking this up @Klaven !

I think, that we are getting a bit off the target here. Currently, the validators have 2 calls to `klog.Errorf` and one call to `klog.Infof` (which, I think, should really be `klog.Errorf`).
In any case, the calls in `package_validator.go` should be turned into `errors.Wrapf` calls.
The call in `kernel_validator.go` should be turned into an `errors.New` and then returned to the caller.

I don't think we should have direct logging from the validators package and the small amount of calls to klog proves this.

rosti commented:
/priority important-longterm

yastij commented:
+1 with @rosti - I still think we should not have logging semantics inside the validators

Commit:
replaced logs in validators to be warnings

Originally kubeadm used klot in the validators to log
These logs are converted to warnings and klog was removed.

Klaven commented:
@rosti @neolit123 @yastij I have updated this. All log messages are errors. please let me know your thoughts!

Klaven commented on cmd/kubeadm/app/util/system/package_validator.go:
```diff
@@ -127,8 +126,7 @@ func (validator *packageValidator) validate(packageSpecs []PackageSpec, manager
 		if err != nil {
-			klog.V(1).Infof("Failed to get the version for the package %q: %s\n", packageName, err)
-			errs = append(errs, err)
+			errs = append(errs, errors.Wrapf(err, "Failed to get the version for the package %q", packageName))
```
I could have just deleted the log as we where already returning the error. but I thought out test made it more clear, what do you think?

rosti commented on cmd/kubeadm/app/util/system/package_validator.go:
```diff
@@ -127,7 +126,6 @@ func (validator *packageValidator) validate(packageSpecs []PackageSpec, manager
 		version, err := manager.getPackageVersion(packageName)
 		if err != nil {
-			klog.V(1).Infof("Failed to get the version for the package %q: %s\n", packageName, err)
 			errs = append(errs, err)
```
Ideally, this should be:
```
errs = append(errs, errors.Wrapf(err, "Failed to get the version of package %q", packageName))
```

Klaven replied:
hahahahaha. So originally I did what you are saying here. BUT the error returned from `manager.getPackageVersion()` effectively says the same exact thing. I caught this because we have a test and the test failed because the error returned changed what it said. and it was mostly redundant. I can change it back to this way and update the test, it just did not seem necessary yesterday.

rosti replied:
Ok, let's leave it like that then.

rosti commented on cmd/kubeadm/app/util/system/package_validator.go:
```diff
@@ -145,7 +143,6 @@ func (validator *packageValidator) validate(packageSpecs []PackageSpec, manager
 		sv, err := semver.Make(toSemVer(version))
 		if err != nil {
-			klog.Errorf("Failed to convert %q to semantic version: %s\n", version, err)
 			errs = append(errs, err)
```
Again:
```
errs = append(errs, errors.Wrapf(err, "Failed to convert %q to semantic version", version))
```

rosti reviewed (changes requested):
Thanks @Klaven !
Overall it's taking shape, but we don't need the warnings. Digging deeper into the only warning use case, it turns out we aren't parsing the kernel config in the same way Kconfig does. Hence, we don't need them.

Klaven commented on cmd/kubeadm/app/util/system/kernel_validator.go:
```diff
@@ -233,34 +235,35 @@ func (k *KernelValidator) getKernelConfigReader() (io.Reader, error) {
 		fields := strings.Split(line, "=")
 		if len(fields) != 2 {
-			klog.Errorf("Unexpected fields number in config %q", line)
+			errs = append(errs, errors.Errorf("Unexpected fields number in config %q", line))
```
@rosti I understand what you are saying. I feel like a change to the logic is outside the scope of this PR... and if I understand correctly this requires a change in the logic. If you don't think it's outside the scope of this PR I can get it done.

rosti replied:
Indeed, that does imply a slight change to the logic. However that is trivial enough and it does remove the need for the klog (hence it gets removed and this is the PR target). Hence, I am all in for this.

neolit123 commented:
btw, we are in code freeze and this can only merge once code thaw begins.

rosti reviewed (commented):
@Klaven , now that CF is over, can you remove the notion of "warning" returns so we can merge this PR?

rosti commented on cmd/kubeadm/app/util/system/kernel_validator.go:
```diff
@@ -233,34 +235,31 @@ func (k *KernelValidator) getKernelConfigReader() (io.Reader, error) {
+		config[fields[0]] = kConfigOption(fields[1][0])
 	}
-	return config, nil
+	return config, errorsutil.NewAggregate(errs), nil
```
To clarify my previous comment, we can remove the warning return here and from the other funcs, since we always return nothing (we never append anything to `errs`).

rosti commented on cmd/kubeadm/app/util/system/kernel_validator_test.go:
```diff
@@ -207,7 +207,8 @@ CONFIG_3=n`
 		Reporter: DefaultReporter,
 	}
-	got, err := v.parseKernelConfig(bytes.NewReader([]byte(config)))
+	got, warn, err := v.parseKernelConfig(bytes.NewReader([]byte(config)))
```
`parseKernelConfig` now returns just a couple of results (`got` and `err`), so this test won't compile.

rosti reviewed (commented):
Thanks @Klaven !
Just a couple of nits and we are there.

rosti reviewed (approved):
Thanks @Klaven !
/approve
/lgtm

yastij reviewed (commented):
/lgtm
